native_c_extension/c_custom_to_s.c

```diff
@@ -0,0 +1,493 @@
+  VALUE ret;
+  // the buffer max size should be "big enough" and it's indeterminate
+  int buffer_max = MAX_STRING_LENGTH * MAX_HASH_ENUMERATION * MAX_EXPECTED_NESTED_LEVELS;
+  char buffer[buffer_max];
+
+  int hash_len = RHASH_SIZE(value);
+  int max_len = hash_len;
+  int offset = 0;
+  int remaining_elements = 0;
+  if (max_len > MAX_HASH_ENUMERATION) {
+    remaining_elements = max_len - MAX_HASH_ENUMERATION;
+    max_len = MAX_HASH_ENUMERATION;
+  }
+
+  snprintf(&buffer[offset], 1 + 1, "%s", "{");
+  offset += 1;
+
+
+  int counter = 0;
+  // pass state from this function into the callback function that
+  // executes during each hash iteration
+  struct method_c_custom_to_s_hash_iterator_state_s state;
+  state.self = self;
+  state.buffer_max = buffer_max;
+  state.buffer = buffer;
+  state.offset = &offset;
+  state.counter = &counter;
+  state.max_len = max_len;
+  state.remaining_elements = remaining_elements;
+  state.remaining_elements_shown = 0;
+  rb_hash_foreach(value, &method_c_custom_to_s_hash_iterator, (VALUE) &state);
+
+  // 2: "}" and \0
+  method_c_custom_to_s_check_buffer_size(*state.offset, 2, state.buffer_max);
+  snprintf(&buffer[offset], 1 + 1, "%s", "}");
+  offset += 1;
+  
+  ret = rb_str_new_cstr(buffer);
+
+  return ret;
+}
+
+/* Returns the bits of an 32-bit unsigned int variable over a given
+ * position. Begin bit is included, but end bit is excluded. */
+static unsigned int get_bits(unsigned int char_reg, int begin, int end)
+{
+    int i = 0;
+    unsigned int bits = 0;
+
+    for (i = begin; i < end; i++)
+    {
+        if (char_reg & (1 << i))
+        {
+            bits |= (1 << (i - begin));
+        }
+    }
+
+    return bits;
+}
+
+/* Counts the number of 8-bit bytes that are required to encode a given
+ * Unicode code point using UTF-8 coding. */
+static int get_num_bytes(unsigned int character)
+{
+    int num_bytes = 0;
+    int i = 0;
+    int bit_pos = 0;
+    const int num_bits = sizeof(unsigned int) * CHAR_BIT;
+
+    for (i = 0; i < num_bits; i++)
+    {
+        if (character & (1 << i))
+        {
+            bit_pos = i;
+        }
+    }
+
+    /* Double-byte encoding (e.g. Scandinavian letters) */
+    if ((bit_pos > 6) && (bit_pos <= 10))
+    {
+        num_bytes = 2;
+    }
+    /* Triple-byte encoding (e.g. CJK characters) */
+    else if ((bit_pos > 10) && (bit_pos <= 15))
+    {
+        num_bytes = 3;
+    }
+    /* Quad-byte encoding (e.g. special characters) */
+    else if (bit_pos > 15)
+    {
+        num_bytes = 4;
+    }
+    /* Single-byte encoding (ASCII characters) */
+    else
+    {
+        num_bytes = 1;
+    }
+
+    return num_bytes;
+}
+
+/*
+ * Encodes the given character's Unicode code point using UTF-8 coding.
+ * Parameters:
+ *      character   Unicode code point as an unsigned integer.
+ *      utf8        Array for UTF-8 encoded bytes (can be 1..4 bytes).
+ */
+int utf8_encode(unsigned int character, unsigned char utf8[])
```

Comment:
I don't understand why you're dealing with encoding at all. Encoding is hard and it will be subtly buggy unless you have years of polishing and bugfixing as I assume Ruby's library has.

Can't you just copy the string and imbue it with the source encoding on creation?

Reply:
I'm dealing with encoding because the existing `custom_display_string` encodes in utf-8. https://github.com/applandinc/appmap-ruby/blob/master/lib/appmap/event.rb#L104.  If strings don't get encoded I see errors like
```
 in `generate': source sequence is illegal/malformed utf-8 (JSON::GeneratorError)
```

I'm dealing with encoding in C because it's 6x-9x faster than encoding in Ruby.  But I didn't get it to work.

Reply:
But can't you just imbue the string with the original string encoding (presumably utf8) and let ruby deal with it? I suppose it would be problematic if you have multiple different encodings going to the same output string, but this should rarely happen (if ever).

Reply:
It should be possible to reuse the encoding of the initial string.  I didn't try this. I'll try.